mypyc/genops.py

```diff
@@ -547,45 +556,31 @@ def f(self, x: object) -> int: ...
                       cls.name, self.module_name,
                       FuncSignature(rt_args, ret_type), blocks, env)
 
-    def contains_func_def(self, stmt: Statement) -> bool:
-        """Traverses a Statement recursively using DFS and checks if it or any nested statements
-        contain a FuncDef.
-        """
-        if isinstance(stmt, (FuncItem, ForStmt, WhileStmt)):
-            if (any(isinstance(s, FuncDef) for s in stmt.body.body) or
-                    any(self.contains_func_def(s) for s in stmt.body.body)):
-                return True
-
-        if isinstance(stmt, IfStmt):
-            for block in stmt.body:
-                if (any(isinstance(s, FuncDef) for s in block.body) or
-                        any(self.contains_func_def(s) for s in block.body)):
-                    return True
-            if stmt.else_body:
-                if (any(isinstance(s, FuncDef) for s in stmt.else_body.body) or
-                        any(self.contains_func_def(s) for s in stmt.else_body.body)):
-                    return True
-
-        return False
-
-    def gen_func_def(self, fdef: FuncDef, sig: FuncSignature,
-                     class_name: Optional[str] = None) -> FuncIR:
+    def gen_func_def(self, fitem: FuncItem, sig: FuncSignature,
+                     class_name: Optional[str] = None) -> FuncInfo:
         """Generates and returns the FuncIR for a given FuncDef.
 
-        If the given FuncDef is a nested function, then we generate a callable class representing
-        the function and use that instead of the actual function. if the given FuncDef contains a
+        If the given FuncItem is a nested function, then we generate a callable class representing
+        the function and use that instead of the actual function. if the given FuncItem contains a
         nested function, then we generate an environment class so that inner nested functions can
-        access the environment of the given FuncDef.
+        access the environment of the given FuncItem.
         """
-        self.fn_info = FuncInfo(fdef, self.gen_func_ns())
+        if isinstance(fitem, LambdaExpr):
+            name = '__mypyc_lambda_{}__'.format(self.lambda_counter)
+        elif isinstance(fitem, FuncDef):
+            name = fitem.name()
+        else:
+            assert False, 'Must be of type LambdaExpr or FuncDef'
```

Comment:
Maybe make the name an argument, instead of casing on it? Callers all know whether it is a Lambda or a FuncDef already.

Reply:
Done.